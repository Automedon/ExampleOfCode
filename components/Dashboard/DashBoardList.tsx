import React from "react";
import { useQuery } from "@apollo/react-hooks";
import { Waypoint } from "react-waypoint";
import withApollo from "../../lib/withApollo";
import { ICar } from "../../server/models/car.model";
import { Field, Form, Formik } from "formik";
import { MenuItem, InputLabel, FormControl, Button } from "@material-ui/core";
import {
  FETCH_SEARCH_DATA_QUERY,
  SEARCH_CARS_QUERY,
} from "../../graphql/query";
import CheckLogin from "../CheckLogin/CheckLogin";
import {
  FormWrapper,
  Row,
  StyledSelect,
  DisplayCarsGrid,
  CarInStock,
} from "./DashboardStyled";

const DashBoardList = () => {
  const searchDataCars = useQuery(FETCH_SEARCH_DATA_QUERY, {
    fetchPolicy: "cache-first",
  });
  const objectSearchDataLoading = searchDataCars.loading;
  const objectSearchData = searchDataCars.data;
  const { data, loading, error, fetchMore, refetch } = useQuery(
    SEARCH_CARS_QUERY,
    {
      variables: {
        carModel: "",
        carMake: "",
        carInStock: "",
        carYear: 0,
        page: 1,
        limit: 20,
      },
      fetchPolicy: "cache-and-network",
    }
  );
  if (error) return <div>Error...</div>;
  if (
    (typeof data === "undefined" && loading) ||
    !data.searchCars ||
    !data.searchCars.docs ||
    objectSearchDataLoading
  ) {
    return <div>Loading...</div>; //TODO spinner
  }
  return (
    <>
      <CheckLogin>
        <Formik
          initialValues={{ make: "", model: "", inStock: "", carYear: 0 }}
          onSubmit={(values, { setSubmitting, resetForm }) => {
            refetch({
              carModel: values.model,
              carMake: values.make,
              carInStock: values.inStock,
              carYear: values.carYear,
              page: 1,
              limit: 20,
            });
            setSubmitting(false);
          }}
        >
          {({
            values,
            handleChange,
            handleBlur,
            handleSubmit,
            isSubmitting,
            setFieldValue,
          }) => (
            <Form onSubmit={handleSubmit}>
              <FormWrapper>
                <Row>
                  <FormControl>
                    <InputLabel id="demo-simple-select-label">Brand</InputLabel>
                    <Field
                      labelId="demo-simple-select-label"
                      as={StyledSelect}
                      type="text"
                      name="make"
                      onChange={(e: React.FormEvent<HTMLFormElement>) => {
                        setFieldValue("model", "");
                        handleChange(e);
                      }}
                      onBlur={handleBlur}
                      value={values.make}
                    >
                      {objectSearchData &&
                        [
                          <MenuItem value={""} key={Math.random()}>
                            All
                          </MenuItem>,
                        ].concat(
                          Object.keys(objectSearchData.fetchCars.myObject).map(
                            (key) => {
                              return (
                                <MenuItem value={key} key={key + Math.random()}>
                                  {key}
                                </MenuItem>
                              );
                            }
                          )
                        )}
                    </Field>
                  </FormControl>
                </Row>
                <Row>
                  <FormControl>
                    <InputLabel id="model-simple-select-label">
                      Model
                    </InputLabel>
                    <Field
                      labelId="model-simple-select-label"
                      as={StyledSelect}
                      type="text"
                      name="model"
                      onChange={handleChange}
                      onBlur={handleBlur}
                      value={values.model}
                    >
                      {[
                        <MenuItem value={""} key={Math.random()}>
                          All
                        </MenuItem>,
                      ].concat(
                        (values.make &&
                          objectSearchData.fetchCars.myObject[
                            values.make
                          ].carModel.map((name: string) => {
                            return (
                              <MenuItem value={name} key={name + Math.random()}>
                                {name}
                              </MenuItem>
                            );
                          })) ||
                          []
                      )}
                    </Field>
                  </FormControl>
                </Row>
                <Row>
                  <FormControl>
                    <InputLabel id="year-simple-select-label">Year</InputLabel>
                    <Field
                      labelId="year-simple-select-label"
                      as={StyledSelect}
                      type="text"
                      name="carYear"
                      onChange={handleChange}
                      onBlur={handleBlur}
                      value={values.carYear}
                    >
                      {[
                        <MenuItem value={0} key={Math.random()}>
                          All
                        </MenuItem>,
                      ].concat(
                        (values.make &&
                          objectSearchData.fetchCars.myObject[
                            values.make
                          ].carYear.map((name: number) => {
                            return (
                              <MenuItem value={name} key={name + Math.random()}>
                                {name}
                              </MenuItem>
                            );
                          })) ||
                          []
                      )}
                    </Field>
                  </FormControl>
                </Row>
                <Row>
                  <FormControl>
                    <InputLabel id="inStock-simple-select-label">
                      Stock
                    </InputLabel>
                    <Field
                      labelId="inStock-simple-select-label"
                      as={StyledSelect}
                      type="text"
                      name="inStock"
                      onChange={handleChange}
                      onBlur={handleBlur}
                      value={values.inStock}
                    >
                      <MenuItem value={""} key={Math.random()}>
                        All
                      </MenuItem>
                      <MenuItem value={"false"} key={Math.random()}>
                        Not In Stock
                      </MenuItem>
                      <MenuItem value={"true"} key={Math.random()}>
                        In Stock
                      </MenuItem>
                    </Field>
                  </FormControl>
                </Row>
                <Row>
                  <Button
                    variant="contained"
                    type="submit"
                    disabled={isSubmitting}
                    style={{
                      margin: "10px 0px",
                    }}
                  >
                    Submit
                  </Button>
                </Row>
              </FormWrapper>
            </Form>
          )}
        </Formik>
      </CheckLogin>
      <div
        style={{
          display: "flex",
          flexDirection: "column",
          alignItems: "center",
        }}
      >
        {data.searchCars.docs.map(
          (
            {
              _id,
              carCost,
              carInStock,
              carImage,
              carMake,
              carModel,
              carVin,
              carYear,
            }: ICar,
            i: number
          ) => {
            return (
              <DisplayCarsGrid key={_id}>
                <img className="carImg" src={carImage} alt={carImage} />
                <div className="carCost">{carCost}</div>
                <div className="cost">Cost </div>
                <CarInStock available={carInStock}>
                  {carInStock ? "Available" : "Not available"}
                </CarInStock>
                <div className="inStock">In Stock </div>
                <div className="carMake">{carMake}</div>
                <div className="make">Brand</div>
                <div className="carModel">{carModel}</div>
                <div className="model">Model </div>
                <div className="carVin">{carVin}</div>
                <div className="vin">Vin Number </div>
                <div className="carYear">{carYear}</div>
                <div className="year">Year </div>
                {i === data.searchCars.docs.length - 10 && (
                  <Waypoint
                    onEnter={() =>
                      data.searchCars.hasNextPage &&
                      !loading &&
                      !objectSearchDataLoading &&
                      fetchMore({
                        variables: {
                          limit: 20,
                          page: data.searchCars.nextPage,
                        },
                        updateQuery: (prev: any, { fetchMoreResult }) => {
                          if (!fetchMoreResult) return prev;
                          return Object.assign({}, prev, {
                            searchCars: {
                              ...prev.searchCars,
                              ...fetchMoreResult.searchCars,
                              docs: [
                                ...prev.searchCars.docs,
                                ...fetchMoreResult.searchCars.docs,
                              ],
                            },
                          });
                        },
                      })
                    }
                  />
                )}
              </DisplayCarsGrid>
            );
          }
        )}
        <h2 style={{ textAlign: "center" }}>Automedon CarYard</h2>
      </div>
    </>
  );
};
export const TestingDashBoardList = DashBoardList;
export default withApollo(DashBoardList);
