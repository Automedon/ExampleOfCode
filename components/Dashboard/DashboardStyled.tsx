import styled from "styled-components";
import Select from "@material-ui/core/Select";

export const StyledSelect = styled(Select)`
  min-width: 300px;
`;

export const DisplayCarsGrid = styled.div`
  display: grid;
  grid-template-rows: repeat(2, 100px);
  grid-template-columns: repeat(8, 100px);
  grid-template-areas:
    "carImg carImg make    model    cost    year    vin    inStock"
    "carImg carImg carMake carModel carCost carYear carVin carInStock";
  align-items: stretch;
  justify-items: stretch;
  text-align: center;
  border: 1px solid black;
  @media (max-width: 700px) {
    max-width: 400px;
    grid-template-rows: repeat(4, 100px);
    grid-template-columns: repeat(4, 79px);
    grid-template-areas:
      "carImg carImg make    model   "
      "carImg carImg carMake carModel"
      "cost    year    vin    inStock"
      "carCost carYear carVin carInStock";
    .carModel {
      grid-area: carVin;
      word-break: break-all;
    }
  }
  * {
    box-sizing: content-box;
    vertical-align: bottom;
    background: azure;
    font-size: 18px;
    border: 1px solid black;
  }
  .carImg {
    grid-area: carImg;
    width: 100%;
    height: 100%;
  }
  .vin {
    grid-area: vin;
  }
  .carVin {
    grid-area: carVin;
    word-break: break-all;
  }
  .cost {
    grid-area: cost;
  }
  .carCost {
    grid-area: carCost;
    font-weight: bolder;
  }
  .inStock {
    grid-area: inStock;
  }
  .make {
    grid-area: make;
  }
  .carMake {
    grid-area: carMake;
    font-weight: bolder;
  }
  .model {
    grid-area: model;
  }
  .carModel {
    grid-area: carModel;
    font-weight: bolder;
  }
  .year {
    grid-area: year;
  }
  .carYear {
    grid-area: carYear;
    font-weight: bolder;
  }
`;
export const FormWrapper = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
`;
export const Row = styled.div`
  display: flex;
  justify-content: flex-start;
  div {
    margin-right: 5px;
  }
`;
export const CarInStock = styled.div<{ available: boolean }>`
  & {
    grid-area: carInStock;
    color: ${(props) => (props.available ? "green" : "red")};
    font-weight: bold;
  }
`;
