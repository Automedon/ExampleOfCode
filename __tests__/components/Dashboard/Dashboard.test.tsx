import React from "react";
import renderer, { act } from "react-test-renderer";
import { MockedProvider } from "@apollo/react-testing";
import wait from "waait";

import "jest-styled-components";

import { TestingDashBoardList } from "../../../components/Dashboard/DashBoardList";
import {
  FETCH_SEARCH_DATA_QUERY,
  SEARCH_CARS_QUERY,
} from "../../../graphql/query";

const mocks = [
  {
    request: {
      query: SEARCH_CARS_QUERY,
      variables: {
        carModel: "",
        carMake: "",
        carInStock: "",
        carYear: 0,
        page: 1,
        limit: 20,
      },
    },
    result: {
      data: {
        searchCars: {
          docs: [
            {
              _id: "1",
              carCost: "$5000",
              carImage: "https://dummyimage.com/600x400/000/fff",
              carInStock: true,
              carMake: "BMW",
              carModel: "x5",
              carVin: "5555555",
              carYear: 2005,
            },
          ],
          hasNextPage: false,
          nextPage: 1,
        },
      },
    },
  },
  {
    request: {
      query: FETCH_SEARCH_DATA_QUERY,
    },
    result: {
      data: {
        fetchCars: {
          myObject: {
            BWM: {
              carModel: ["x5", "x7"],
              carYear: [2000, 2019],
            },
          },
          myValue: "JSON",
        },
      },
    },
  },
];

it("DashBoardList renders", async () => {
  const tree = renderer.create(
    <MockedProvider mocks={mocks} addTypename={false}>
      <TestingDashBoardList />
    </MockedProvider>
  );
  await act(() => wait(0));
  expect(tree).toMatchSnapshot();
});
