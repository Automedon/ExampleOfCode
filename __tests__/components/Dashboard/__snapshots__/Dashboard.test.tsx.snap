// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DashBoardList renders 1`] = `
Array [
  <div
    style={
      Object {
        "textAlign": "center",
      }
    }
  >
    <div>
      <a
        href="/login"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        Go to Login
      </a>
    </div>
    ***Here is a content you need to be logged to see it***
  </div>,
  <div
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "column",
      }
    }
  >
    <div
      className=""
    >
      <img
        alt="https://dummyimage.com/600x400/000/fff"
        className="carImg"
        src="https://dummyimage.com/600x400/000/fff"
      />
      <div
        className="carCost"
      >
        $5000
      </div>
      <div
        className="cost"
      >
        Cost 
      </div>
      <div
        className=""
      >
        Available
      </div>
      <div
        className="inStock"
      >
        In Stock 
      </div>
      <div
        className="carMake"
      >
        BMW
      </div>
      <div
        className="make"
      >
        Brand
      </div>
      <div
        className="carModel"
      >
        x5
      </div>
      <div
        className="model"
      >
        Model 
      </div>
      <div
        className="carVin"
      >
        5555555
      </div>
      <div
        className="vin"
      >
        Vin Number 
      </div>
      <div
        className="carYear"
      >
        2005
      </div>
      <div
        className="year"
      >
        Year 
      </div>
    </div>
    <h2
      style={
        Object {
          "textAlign": "center",
        }
      }
    >
      Automedon CarYard
    </h2>
  </div>,
]
`;
